docs/src/best-practices-js.md

```diff
@@ -0,0 +1,200 @@
+
+### Testing with a database
+
+If working with a database then make sure you control the data. Test against a staging environment and make sure it doesn't change. For visual regression tests make sure the operating system and browser versions are the same.
+
+## Best Practices
+
+### Use locators
+
+In order to write end to end tests we need to first find elements on the webpage. We can do this by using Playwright's built in [locators](./locators.md). Locators come with auto waiting and retry-ability. Auto waiting means that Playwright performs a range of actionability checks on the elements, such as ensuring the element is visible and enabled before it performs the click. To make tests resilient, we recommend prioritizing user-facing attributes and explicit contracts.
+
+```js
+👍 page.getByRole('button', { name: 'submit' })
+```
+
+#### Prefer user-facing attributes to xpath or css selectors
+
+Your DOM can easily change so having your tests depend on your DOM structure can lead to failing tests. For example consider selecting this button by its CSS classes. Should the designer change something then the class might change breaking your test. 
+
+
+```js
+👎 page.locator('button.buttonIcon.episode-actions-later')
+```
+
+Use locators that are resilient to changes in the DOM.
+
+```js
+👍 page.getByRole('button', { name: 'submit' })
+```
+### Generate locators
+
+Playwright has a [test generator](./codegen.md) that can generate tests and pick locators for you. It will look at your page and figure out the best locator, prioritizing role, text and test id locators. If the generator finds multiple elements matching the locator, it will improve the locator to make it resilient and uniquely identify the target element, so you don't have to worry about failing tests due to locators.
+
+#### Use codegen to generate locators
+
+To pick a locator you can run the codegen command and click on the pick locator button. Then hover over your page in the browser window and click on the element you want to pick. You can then copy and paste this locator into your code. You can also use the codegen to record a test for you.
+
+```bash
+npx playwright codegen
+```
+#### Use the VS Code extension to generate locators
+
+You can also use the [VS Code Extension](./getting-started-vscode.md) to generate locators as well as record a test. The VS Code extension also gives you a great developer experience when writing, running and debugging tests.
+
+### Use web first assertions
+
+Assertions are a way to verify that the expected result and the actual result matched or not. By using [web first assertions](./test-assertions.md) Playwright will wait until the expected condition is met. For example, when testing a toast message, a test would click a button that makes a toast message appear and check that the toast message is there. If the toast takes half a second to appear, assertions such as `toBeVisible()` will wait and retry if needed.
```

Comment:
Please consider making "toast" be a link to a good definition of what a toast message is. I had never heard that phrase before, so it's probably new to others as well.